tensorcircuit/shadows.py

```diff
@@ -0,0 +1,282 @@
+    vvv = backend.vmap(vv, vectorized_argnums=(0, 1))
+
+    return vvv(pauli_strings, snapshots)
+
+
+def shadow_state(snapshots, pauli_strings, sub: Optional[Sequence[int]] = None):
+    '''
+        ns: number of snapshots
+        nq: number of qubits
+        pauli_strings = (ns, nq) or (ns, repeat, nq)
+        snapshots = (ns, repeat, nq)
+
+        return:
+        shadow_state = (2 ** nq, 2 ** nq)
+    '''
+    return backend.mean(global_shadow_states(snapshot_states(snapshots, pauli_strings), sub), axis=(0, 1))
+
+
+def expection_ps_shadow(snapshots, pauli_strings, x: Optional[Sequence[int]] = None, y: Optional[Sequence[int]] = None,
+                   z: Optional[Sequence[int]] = None, ps: Optional[Sequence[int]] = None, k: int = 1):
+    '''
+        ns: number of snapshots
+        nq: number of qubits
+        pauli_strings = (ns, nq) or (ns, repeat, nq)
+        snapshots = (ns, repeat, nq)
+
+        return:
+        expection = (1,)
+    '''
+    ss_states = snapshot_states(snapshots, pauli_strings)  # (ns, repeat, nq, 2, 2)
+    ns, repeat, nq, _, _ = ss_states.shape
+    ns *= repeat
+    ss_states = backend.reshape(ss_states, (ns, nq, 2, 2))
+
+    if ps is not None:
+        ps = np.array(ps)  # (nq,)
+    else:
+        ps = np.zeros(nq, dtype=int)
+        if x is not None:
+            for i in x:
+                ps[i] = 0
+        if y is not None:
+            for i in y:
+                ps[i] = 1
+        if z is not None:
+            for i in z:
+                ps[i] = 2
+
+    paulis = backend.convert_to_tensor(
+        backend.cast(np.array([[[1, 0], [0, 1]], [[0, 1], [1, 0]], [[0, -1j], [1j, 0]], [[1, 0], [0, -1]]]),
+                     dtype=ss_states.dtype))
+
+    def sqp(dm, p_idx):
+        return 3 * backend.real(backend.trace(paulis[p_idx] @ dm))
+
+    v = backend.vmap(sqp, vectorized_argnums=(0, 1))  # (nq,)
+
+    def prod(dm, p_idx):
+        tensor = v(dm, p_idx)
+        return backend.shape_prod(tensor)
+
+    vv = backend.vmap(prod, vectorized_argnums=0)  # (ns,)
+
+    batch = ns // k
+    means = []
+    for i in range(0, ns, batch):
+        means.append(backend.mean(vv(ss_states[i: i + batch], ps)))
+    return means
+
+
+def entropy_shadow(snapshots, pauli_strings, sub: Optional[Sequence[int]] = None, alpha: int = 1):
+    '''
+            ns: number of snapshots
+            nq: number of qubits
+            pauli_strings = (ns, nq) or (ns, repeat, nq)
+            snapshots = (ns, repeat, nq)
+
+            return:
+            entropy = (1,)
+        '''
+    if alpha <= 0:
+        raise ValueError("Alpha should not be less than 1!")
+
+    sdw_rdm = shadow_state(snapshots, pauli_strings, sub)   # (2 ** nq, 2 ** nq)
+
+    evs = backend.relu(backend.eigvalsh(sdw_rdm))
+    if alpha == 1:
+        return -backend.sum(evs * backend.relu(backend.log(evs + 1e-15)))
+    else:
+        return backend.log(backend.sum(backend.power(evs, alpha))) / (1 - alpha)
```

Comment:
renyi entropy could be reconstrcut without reconstructing the global state explicitly? which might be more efficient